src/nomad_measurements/utils.py

```diff
@@ -153,3 +160,304 @@ def get_bounding_range_2d(ax1, ax2):
+        view of the data in addition to storing array data.
+        """
+        if self.data_file.endswith('.h5'):
+            self.data_file = self.data_file.replace('.h5', '.nxs')
+        raise NotImplementedError('Method `write_nx_file` is not implemented.')
+        # TODO add archive data to `hdf5_data_dict` before creating the nexus file. Use
+        # `populate_hdf5_data_dict` method for each quantity that is needed in .nxs
+        # file. Create a NeXus file with the data in `hdf5_data_dict`.
+        # One issue here is as we populate the `hdf5_data_dict` with the archive data,
+        # we will always have to over write the nexus file
+
+    def _write_hdf5_file(self):  # noqa: PLR0912
+        """
+        Method for creating an HDF5 file.
+        """
+        if self.data_file.endswith('.nxs'):
+            self.data_file = self.data_file.replace('.nxs', '.h5')
+        if not self.hdf5_datasets and not self.hdf5_attributes:
+            return
+        # remove the nexus annotations from the dataset paths if any
+        tmp_dict = {}
+        for key, value in self.hdf5_datasets.items():
+            new_key = self._remove_nexus_annotations(key)
+            tmp_dict[new_key] = value
+            tmp_dict[new_key]['hdf5_path'] = self._remove_nexus_annotations(
+                value['hdf5_path']
+            )
+        self.hdf5_datasets = tmp_dict
+        tmp_dict = {}
+        for key, value in self.hdf5_attributes.items():
+            tmp_dict[self._remove_nexus_annotations(key)] = value
+        self.hdf5_attributes = tmp_dict
+
+        # create the HDF5 file
+        with self.archive.m_context.raw_file(self.data_file, 'a') as h5file:
+            h5 = h5py.File(h5file.name, 'a')
+            for key, value in self.hdf5_datasets.items():
+                if value['data'] is None:
+                    self.logger.warning(f'No data found for "{key}". Skipping.')
+                    continue
+                elif value['internal_reference']:
+                    # resolve the internal reference
+                    try:
+                        data = h5[self._remove_nexus_annotations(value['data'])]
+                    except KeyError:
+                        self.logger.warning(
+                            f'Internal reference "{data}" not found. Skipping.'
+                        )
+                        continue
+                else:
+                    data = value['data']
+
+                group_name, dataset_name = key.rsplit('/', 1)
+                group = h5.require_group(group_name)
+
+                if key in h5:
+                    group[dataset_name][...] = data
+                else:
+                    group.create_dataset(
+                        name=dataset_name,
+                        data=data,
+                    )
+                group[dataset_name].attrs.update(value['attrs'])
+                if value['archive_path'] is not None:
```

Comment:
this seems to mix the assignment of the reference to the corresponding archive quantity, which is beyond the scope of the function? There is no way the user can control this, maybe make it an external function?

Reply:
I see, I can add this check inside the static method instead